rpc/dispatch_test.go

```diff
@@ -64,54 +80,106 @@ func websocketHandler(f func(*websocket.Conn)) http.Handler {
 	})
 }
 
-func (s *dispatchSuite) SetUpTest(c *gc.C) {
-	s.BaseSuite.SetUpTest(c)
-	loggo.GetLogger("juju.rpc").SetLogLevel(loggo.TRACE)
-}
-
 func (s *dispatchSuite) TestWSWithoutParamsV0(c *gc.C) {
-	resp := s.request(c, `{"RequestId":1,"Type": "DispatchDummy","Id": "without","Request":"DoSomething"}`)
-	s.assertResponse(c, resp, `{"RequestId":1,"Response":{}}`)
+	err := s.requestV0(c, `{"RequestId":1,"Type": "DispatchDummy","Id": "without","Request":"DoSomething"}`)
+	c.Assert(errors.Is(err, errors.NotSupported), jc.IsTrue)
 }
 
 func (s *dispatchSuite) TestWSWithParamsV0(c *gc.C) {
-	resp := s.request(c, `{"RequestId":2,"Type": "DispatchDummy","Id": "with","Request":"DoSomething", "Params": {}}`)
-	s.assertResponse(c, resp, `{"RequestId":2,"Response":{}}`)
+	err := s.requestV0(c, `{"RequestId":2,"Type": "DispatchDummy","Id": "with","Request":"DoSomething", "Params": {}}`)
+	c.Assert(errors.Is(err, errors.NotSupported), jc.IsTrue)
 }
 
 func (s *dispatchSuite) TestWSWithoutParamsV1(c *gc.C) {
-	resp := s.request(c, `{"request-id":1,"type": "DispatchDummy","id": "without","request":"DoSomething"}`)
+	resp := s.requestV1(c, `{"request-id":1,"type": "DispatchDummy","id": "without","request":"DoSomething"}`)
 	s.assertResponse(c, resp, `{"request-id":1,"response":{}}`)
+	s.ensureFinish(c)
 }
 
 func (s *dispatchSuite) TestWSWithParamsV1(c *gc.C) {
-	resp := s.request(c, `{"request-id":2,"type": "DispatchDummy","id": "with","request":"DoSomething", "params": {}}`)
+	resp := s.requestV1(c, `{"request-id":2,"type": "DispatchDummy","id": "with","request":"DoSomething", "params": {}}`)
 	s.assertResponse(c, resp, `{"request-id":2,"response":{}}`)
+	s.ensureFinish(c)
 }
 
 func (s *dispatchSuite) assertResponse(c *gc.C, obtained, expected string) {
 	c.Assert(obtained, gc.Equals, expected+"\n")
 }
 
+func (s *dispatchSuite) ensureFinish(c *gc.C) {
+	// Ensure the test is finished, before starting another one. This is to
+	// prevent a data race on the dead channel in SetupTest.
+	select {
+	case <-s.dead:
+	case <-time.After(testing.LongWait):
+		c.Fatalf("timeout waiting for response")
+	}
+}
+
+// request performs one request to the test server via websockets.
+func (s *dispatchSuite) requestV0(c *gc.C, req string) error {
+	ws := s.request(c, req)
+
+	result := make(chan struct{})
+
+	go func() {
+		_, _, err := ws.ReadMessage()
+		c.Check(err, gc.NotNil)
+
+		close(result)
+	}()
+
+	select {
+	case err := <-s.dead:
+		return err
```

Comment:
This is potentially problematic as  both requestV0 and ensureFinished are going to be waiting on the dead channel, which will only ever send 1 error, and won't be close() until the test actually exits.

Reply:
I've refactored this completely in a different PR that makes this go away